src/sentry/snuba/metrics/utils.py

```diff
@@ -0,0 +1,143 @@
+__all__ = (
+    "MAX_POINTS",
+    "GRANULARITY",
+    "TS_COL_QUERY",
+    "TS_COL_GROUP",
+    "FIELD_REGEX",
+    "TAG_REGEX",
+    "MetricOperation",
+    "MetricUnit",
+    "MetricType",
+    "OP_TO_SNUBA_FUNCTION",
+    "AVAILABLE_OPERATIONS",
+    "OPERATIONS_TO_ENTITY",
+    "METRIC_TYPE_TO_ENTITY",
+    "ALLOWED_GROUPBY_COLUMNS",
+    "Tag",
+    "TagValue",
+    "MetricMeta",
+    "MetricMetaWithTagKeys",
+    "OPERATIONS",
+    "DEFAULT_AGGREGATES",
+    "UNIT_TO_TYPE",
+    "DerivedMetricParseException",
+    "TimeRange",
+)
+
+
+import re
+from datetime import datetime
+from typing import Collection, Literal, Mapping, Optional, Protocol, Sequence, TypedDict
+
+from sentry.snuba.dataset import EntityKey
+
+MAX_POINTS = 10000
+GRANULARITY = 24 * 60 * 60
+TS_COL_QUERY = "timestamp"
+TS_COL_GROUP = "bucketed_time"
+
+#: Max number of data points per time series:
+# ToDo modify this regex to only support the operations provided
+FIELD_REGEX = re.compile(r"^(\w+)\(((\w|\.|_)+)\)$")
+TAG_REGEX = re.compile(r"^(\w|\.|_)+$")
+
+#: A function that can be applied to a metric
+MetricOperation = Literal["avg", "count", "max", "min", "p50", "p75", "p90", "p95", "p99"]
+MetricUnit = Literal["seconds"]
+#: The type of metric, which determines the snuba entity to query
+MetricType = Literal["counter", "set", "distribution"]
+
+OP_TO_SNUBA_FUNCTION = {
+    "metrics_counters": {"sum": "sumIf"},
+    "metrics_distributions": {
+        "avg": "avgIf",
+        "count": "countIf",
+        "max": "maxIf",
+        "min": "minIf",
+        # TODO: Would be nice to use `quantile(0.50)` (singular) here, but snuba responds with an error
+        "p50": "quantilesIf(0.50)",
+        "p75": "quantilesIf(0.75)",
+        "p90": "quantilesIf(0.90)",
+        "p95": "quantilesIf(0.95)",
+        "p99": "quantilesIf(0.99)",
+    },
+    "metrics_sets": {"count_unique": "uniqIf"},
+}
+
+
+AVAILABLE_OPERATIONS = {
+    type_: sorted(mapping.keys()) for type_, mapping in OP_TO_SNUBA_FUNCTION.items()
+}
```

Comment:
`AVAILABLE_OPERATIONS` is defined on the entity and snuba operations right now, but I think it should rather be defined on the metric type. We may have aggregation functions that aren't directly mapped to snuba (think the `rate` function on counters), and then there's calculated derived metrics.